### PR TITLE
feature: Adding support in HuggingFace estimator for Training Compiler enhanced PyTorch 1.11

### PR DESCRIPTION
**Pull request by Lokiiiiii**
*Description of changes:*
1. Adding new DLC for trcomp-hf-pt 1.11
2. Adding support for new distribution mechanism 'pytorch_xla' to be used only with trcomp-hf-pt 1.11
3. Invalidating heterogeneous clusters with Training Compiler
4. Adding support for G5 instances to be used with Training Compiler

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 1]
```diff
@@ -24,7 +26,7 @@
 class TrainingCompilerConfig(BaseConfig):
     """The SageMaker Training Compiler configuration class."""
 
-    SUPPORTED_INSTANCE_CLASS_PREFIXES = ["p3", "g4dn", "p4"]
+    SUPPORTED_INSTANCE_CLASS_PREFIXES = ["p3", "g4dn", "p4d", "g5"]
```
Comment (pinaraws) [inline src/sagemaker/huggingface/training_compiler/config.py:29]: did you mean to remove p4? is it intentional?

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 2]
```diff
@@ -105,3 +107,40 @@ def validate(
...
+                            "Distribution mechanism 'pytorch_xla' is currently only supported for "
+                            "PyTorch >= 1.11 when Training Compiler is enabled. Received "
+                            "pytorch_version={} which is unsupported."
+                        )
+                        raise ValueError(error_helper_string.format(estimator.pytorch_version))
+            if not pt_xla_present:
+                if estimator.pytorch_version:
+                    if Version(estimator.pytorch_version) in SpecifierSet(">= 1.11"):
+                        error_helper_string = (
+                            "'pytorch_xla' is the only distribution mechanism currently supported "
```
Comment (pinaraws) [inline src/sagemaker/huggingface/training_compiler/config.py:134]: you can still run without specifying distribution, correct?

File: src/sagemaker/training_compiler/config.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ class TrainingCompilerConfig(object):
     """The SageMaker Training Compiler configuration class."""
 
     DEBUG_PATH = "/opt/ml/output/data/compiler/"
-    SUPPORTED_INSTANCE_CLASS_PREFIXES = ["p3", "g4dn", "p4"]
+    SUPPORTED_INSTANCE_CLASS_PREFIXES = ["p3", "g4dn", "p4d", "g5"]
```
Comment (pinaraws) [inline src/sagemaker/training_compiler/config.py:24]: same here, is p4 removed intentionally?

File: tests/integ/test_training_compiler.py [hunk 4]
```diff
@@ -105,10 +119,10 @@ def test_huggingface_pytorch(
                 "per_device_train_batch_size": 128,
                 "output_dir": "/opt/ml/model",
             },
-            environment={"GPU_NUM_DEVICES": "1"},
```
Comment (pinaraws) [inline tests/integ/test_training_compiler.py:108]: who is going to set this now?

Comment (Lokiiiiii) [reply tests/integ/test_training_compiler.py:108]: Training Toolkit will set this for Distributed Training.
Pytorch_Xla init will default this to 1 for all other cases.

File: src/sagemaker/huggingface/training_compiler/config.py (hunk 1 above)
Comment (Lokiiiiii) [reply src/sagemaker/huggingface/training_compiler/config.py:29]: `p4d` is the only available flavor of `p4` AFAIK. Just making the check more specific.

File: src/sagemaker/huggingface/training_compiler/config.py (hunk 2 above)
Comment (Lokiiiiii) [reply src/sagemaker/huggingface/training_compiler/config.py:134]: Yes, this is just blocking customers from using other distribution mechanisms like mpi or pytorch ddp.

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 5]
```diff
@@ -85,7 +87,7 @@ def validate(
         """Checks if SageMaker Training Compiler is configured correctly.
 
         Args:
-            estimator (str): A estimator object
+            estimator (sagemaker.huggingface.HuggingFace): A estimator object
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/training_compiler/config.py:90]:
```suggestion
            estimator (:class:`sagemaker.huggingface.HuggingFace`): An estimator object.
```

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 6]
```diff
@@ -105,3 +107,40 @@ def validate(
...
             raise ValueError(error_helper_string)
+
+        if estimator.distribution:
+            pt_xla_present = "pytorch_xla" in estimator.distribution
+            pt_xla_enabled = estimator.distribution.get("pytorch_xla", {}).get("enabled", False)
+            if pt_xla_enabled:
+                if estimator.tensorflow_version:
+                    error_helper_string = (
+                        "Distribution mechanism 'pytorch_xla' is currently only supported for "
+                        "PyTorch >= 1.11 when Training Compiler is enabled. Received "
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/training_compiler/config.py:118]: What is the word "Received" for?

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 7]
```diff
@@ -105,3 +107,40 @@ def validate(
...
+                            "distribution={} which is unsupported."
+                        )
+                        raise ValueError(error_helper_string.format(estimator.distribution))
+        elif estimator.instance_count and estimator.instance_count > 1:
+            if estimator.pytorch_version:
+                if Version(estimator.pytorch_version) in SpecifierSet(">= 1.11"):
+                    logger.warning(
+                        "Consider setting 'distribution' to 'pytorch_xla' for distributed "
+                        "training with PyTorch >= 1.11 with Training Compiler enabled. This "
+                        "will become the default configuration in the future."
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/training_compiler/config.py:145]:
```suggestion
```

Comment (mchoi8739) [reply src/sagemaker/huggingface/training_compiler/config.py:145]: We never publicly document any future plans

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 8]
```diff
@@ -105,3 +107,40 @@ def validate(
...
+                            "for PyTorch >= 1.11 when Training Compiler is enabled. Received "
+                            "distribution={} which is unsupported."
+                        )
+                        raise ValueError(error_helper_string.format(estimator.distribution))
+        elif estimator.instance_count and estimator.instance_count > 1:
+            if estimator.pytorch_version:
+                if Version(estimator.pytorch_version) in SpecifierSet(">= 1.11"):
+                    logger.warning(
+                        "Consider setting 'distribution' to 'pytorch_xla' for distributed "
+                        "training with PyTorch >= 1.11 with Training Compiler enabled. This "
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/training_compiler/config.py:144]:
```suggestion
                        "training with PyTorch >= 1.11 and SageMaker Training Compiler enabled."
```

File: src/sagemaker/training_compiler/config.py [hunk 9]
```diff
@@ -123,7 +123,7 @@ def validate(
         """Checks if SageMaker Training Compiler is configured correctly.
 
         Args:
-            estimator (str): A estimator object
+            estimator (sagemaker.Estimator): A estimator object
```
Comment (mchoi8739) [inline src/sagemaker/training_compiler/config.py:126]:
```suggestion
            estimator (:class:`sagemaker.estimator.Estimator`): An estimator object.
```

File: src/sagemaker/huggingface/training_compiler/config.py (hunk 6 above)
Comment (Lokiiiiii) [reply src/sagemaker/huggingface/training_compiler/config.py:118]: It's continued in the next line: `Received tensorflow_version={} which is unsupported`

File: src/sagemaker/huggingface/estimator.py [hunk 10]
```diff
@@ -141,6 +141,16 @@ def __init__(
                             }
                         }
                     }
+
+                To enable distributed training with Training Compiler for PyTorch:
+
+                .. code:: python
+
+                    {
+                        "pytorch_xla": {
```
Comment (pinaraws) [inline src/sagemaker/huggingface/estimator.py:150]: Let's make this `pytorchxla` to be consistent with:
`distribution = {"pytorchddp": {"enabled": "true"}}`

**Review by pinaraws — APPROVED** [review]

File: src/sagemaker/huggingface/estimator.py [hunk 11]
```diff
@@ -141,6 +141,16 @@ def __init__(
                             }
                         }
                     }
+
+                To enable distributed training with Training Compiler for PyTorch:
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/estimator.py:145]:
```suggestion
                To enable distributed training with `SageMaker Training Compiler <https://docs.aws.amazon.com/sagemaker/latest/dg/training-compiler.html>`_ for PyTorch:
```

File: src/sagemaker/huggingface/estimator.py [hunk 12]
```diff
@@ -204,6 +214,11 @@ def __init__(
                 raise ValueError(error_string)
             if compiler_config:
                 compiler_config.validate(self)
+        elif distribution is not None and "pytorchxla" in distribution:
+            raise ValueError(
+                "Distributed training through PyTorch XLA is currently only supported "
+                "when Training Compiler is enabled."
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/estimator.py:220]:
```suggestion
                "when SageMaker Training Compiler is enabled. To learn more, see Enable SageMaker Training Compiler at https://docs.aws.amazon.com/sagemaker/latest/dg/training-compiler-enable.html."
```

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 13]
```diff
@@ -105,3 +107,39 @@ def validate(
...
             raise ValueError(error_helper_string)
+
+        if estimator.distribution:
+            pt_xla_present = "pytorchxla" in estimator.distribution
+            pt_xla_enabled = estimator.distribution.get("pytorchxla", {}).get("enabled", False)
+            if pt_xla_enabled:
+                if estimator.tensorflow_version:
+                    error_helper_string = (
+                        "Distribution mechanism 'pytorchxla' is currently only supported for "
+                        "PyTorch >= 1.11 when Training Compiler is enabled. Received "
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/training_compiler/config.py:118]:
```suggestion
                        "PyTorch >= 1.11 when SageMaker Training Compiler is enabled. Received "
```

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 14]
```diff
@@ -105,3 +107,39 @@ def validate(
...
+                        "tensorflow_version={} which is unsupported."
+                    )
+                    raise ValueError(error_helper_string.format(estimator.tensorflow_version))
+                elif estimator.pytorch_version:
+                    if Version(estimator.pytorch_version) in SpecifierSet("< 1.11"):
+                        error_helper_string = (
+                            "Distribution mechanism 'pytorchxla' is currently only supported for "
+                            "PyTorch >= 1.11 when Training Compiler is enabled. Received "
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/training_compiler/config.py:126]:
```suggestion
                            "PyTorch >= 1.11 when SageMaker Training Compiler is enabled. Received "
```

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 15]
```diff
@@ -105,3 +107,39 @@ def validate(
...
+                            "pytorch_version={} which is unsupported."
+                        )
+                        raise ValueError(error_helper_string.format(estimator.pytorch_version))
+            if not pt_xla_present:
+                if estimator.pytorch_version:
+                    if Version(estimator.pytorch_version) in SpecifierSet(">= 1.11"):
+                        error_helper_string = (
+                            "'pytorchxla' is the only distribution mechanism currently supported "
+                            "for PyTorch >= 1.11 when Training Compiler is enabled. Received "
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/training_compiler/config.py:135]:
```suggestion
                            "for PyTorch >= 1.11 when SageMaker Training Compiler is enabled. Received "
```

File: src/sagemaker/huggingface/estimator.py [hunk 16]
```diff
@@ -141,6 +141,16 @@ def __init__(
...
+
+                To enable distributed training with Training Compiler for PyTorch:
+
+                .. code:: python
+
+                    {
+                        "pytorchxla": {
+                            "enabled": True
+                        }
+                    }
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/estimator.py:153]:
```suggestion
                    }
                    
                To learn more, see `SageMaker Training Compiler <https://docs.aws.amazon.com/sagemaker/latest/dg/training-compiler.html>`_ in the *Amazon SageMaker Developer Guide*.
                
                .. note::
                
                    When you use this PyTorch XLA option for distributed training strategy, you must add the ``compiler_config`` parameter and activate SageMaker Training Compiler.
```

**Review by mchoi8739 — CHANGES_REQUESTED** [review]

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run all

**Review by mufaddal-rohawala — APPROVED** [review]

File: src/sagemaker/training_compiler/config.py [hunk 17]
```diff
@@ -132,31 +132,34 @@ def validate(
...
+                requested_instance_class = estimator.instance_type.split(".")[
+                    1
+                ]  # Expecting ml.class.size
+                if not any(
+                    [
+                        requested_instance_class.startswith(i)
+                        for i in cls.SUPPORTED_INSTANCE_CLASS_PREFIXES
+                    ]
+                ):
+                    error_helper_string = (
```
Comment (harishneit) [inline src/sagemaker/training_compiler/config.py:146]: can we add a warning/info logging to inform customers that `p3dn.24xlarge` & `p4d.24xlarge` instances offer the best performance for distributed training?

Comment (Lokiiiiii) [reply src/sagemaker/training_compiler/config.py:146]: Do we just want to generalize to EFA enabled instances or what is special about these 2 instance types ?

File: src/sagemaker/huggingface/training_compiler/config.py [hunk 18]
```diff
@@ -24,7 +26,14 @@
 class TrainingCompilerConfig(BaseConfig):
     """The SageMaker Training Compiler configuration class."""
 
-    SUPPORTED_INSTANCE_CLASS_PREFIXES = ["p3", "g4dn", "p4"]
+    SUPPORTED_INSTANCE_CLASS_PREFIXES = ["p3", "g4dn", "p4d", "g5"]
+    SUPPORTED_INSTANCE_TYPES_WITH_EFA = [
+        "ml.g4dn.8xlarge",
```
Comment (harishneit) [inline src/sagemaker/huggingface/training_compiler/config.py:31]: current EFA is only exposed with `ml.p4d.24xlarge` & `ml.p3dn.24xlarge` for GPU instances, can we remove other GPU instances please. Please refer to https://aws.amazon.com/about-aws/whats-new/2021/05/amazon-sagemaker-supports-elastic-fabric-adapter-distributed-training/ for details.

**Review by harishneit — APPROVED** [review]

File: src/sagemaker/huggingface/estimator.py [hunk 19]
```diff
@@ -141,6 +141,28 @@ def __init__(
                             }
                         }
                     }
+
+                To enable distributed training with
+                `SageMaker Training Compiler <https://docs.aws.amazon.com/sagemaker/latest/dg/training-compiler.html>`_
+                for PyTorch:
```
Comment (mchoi8739) [inline src/sagemaker/huggingface/estimator.py:147]:
```suggestion
                for Hugging Face Transformers with PyTorch:
```

File: src/sagemaker/training_compiler/config.py [hunk 20]
```diff
@@ -132,31 +132,34 @@ def validate(
...
+                        "SageMaker Training Compiler only supports {}"
+                    )
+                    error_helper_string = error_helper_string.format(
+                        requested_instance_class, cls.SUPPORTED_INSTANCE_CLASS_PREFIXES
+                    )
+                    raise ValueError(error_helper_string)
+            elif estimator.instance_type == "local":
                 error_helper_string = (
-                    "Unsupported Instance class {}. SageMaker Training Compiler only supports {}"
+                    "The local mode is not supported by SageMaker Training Compiler."
```
Comment (mchoi8739) [inline src/sagemaker/training_compiler/config.py:156]:
```suggestion
                    "SageMaker Training Compiler doesn't support local mode."
```

**Review by mchoi8739 — APPROVED** [review]

File: src/sagemaker/huggingface/training_compiler/config.py (hunk 18 above)
Comment (Lokiiiiii) [reply src/sagemaker/huggingface/training_compiler/config.py:31]: SMs contract is that if customer can use g5.48xl on EC2 with EFA for non-managed training experience, then a customer should be able to use same hardware features on SM with the managed experience.

File: src/sagemaker/training_compiler/config.py (hunk 17 above)
Comment (harishneit) [reply src/sagemaker/training_compiler/config.py:146]: we want to start with recommending EFA instances we have benchmarked, we can add other instances to the list once we benchmark them.

**Pull request merged**